Automattic/themes
# feat(tools): automate theme changelogs and versioning

#### Changes proposed in this Pull Request:
This PR is a proposal to start using [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) in our repo, in order to use the format as a way to automate changelog creation and version bumping. It uses [Lerna](https://github.com/lerna/lerna) to manage all of the themes as packages, and [commitlint](https://github.com/conventional-changelog/commitlint) to ensure all our commits follow the correct format.

### Integration with deploy script
We should integrate this with our current deploy script. My thinking at the moment is that we should do something like this:
- Versioning and changelog updates happen on PR merge via GitHub action (as per changes in this PR)
- Deploy script doesn’t handle version bumps but continues as is otherwise

Or...

- Deploy script calls the new command in this PR (`update-themes`) instead of the current version bump scripts (we'd need to use conventional commits on all commits in this case I think, as it skips the GH action on PR merge)

### Changes to our process
We'd need to start using conventional commits, but we can customise the format if we want to. These are used to determine which changelogs are updated, what’s included in the changelogs, and to facilitate the SemVer version bumping. It's also useful for consistently highlighting breaking changes.

It's actually only the PR title that needs to follow the conventional commit format, so we could just use it there rather than all commits if preferred.

**How this will work with the new GitHub action:**
- Start a new branch and ideally commit all changes using `npm run commit`, to catch breaking changes and correct labeling (chore, feat, docs, etc). We can commit manually rather than using this command if preferred.
- Create a PR
- Squash and merge
- Make sure PR title follows conventional commit format (PR title will be linted to ensure it follows this format using [this action](https://github.com/marketplace/actions/semantic-pull-request))
- GitHub action should automatically update changelogs and version bump the themes relating to the changeset in a new commit (the commit will be attributed to the Github actions bot)

### Testing

**To test locally:**
- In `lerna.json`, remove `"allowBranch": "trunk"` in order to test on this branch
- Make a change to a theme (e.g. change a readme file)
- Commit these changes following the conventional commit format (e.g. ‘docs(blockbase): update readme’). There's a new command for this: `npm run commit`, which walks you through the format (instead of needing to remember it)
- Run `npm run update-themes`
- This should update the changed themes with a new version number and an updated changelog, and then create a new commit with these changes

### Questions
- Should we use conventional commits on all commits or just for PR titles? I think I'm leaning towards just using it in PR titles, that way we can remove commitlint and commitizen, and only lint the PR titles. We could add extra instructions to our PR templates to help out with title naming. This also feels like a less significant change to our workflow.
- Consider other ways we could use other Lerna features if we went down this route (managing the link between parent and child themes, running scripts across multiple themes)
